_posts/2019-03-26-k8s-network.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: post
+title:  "br1dging the Gap I: Working with tap devices in k8s"
+date:   2019-03-26
+tags: solo5 rumprun kubernetes cni oci
+author: Brandon Lum
+---
+
+
+This is the first of several blog posts in the "`br1`dging the Gap" series! The series of posts will talk about the various gaps between what containers expect (and how the orchestration system provides), and what nabla containers expect. This was a prominent theme in the writing of runnc, and we hope to be able to share a little about our design choices in runnc. 
+These set of posts relate to Nabla containers as well as other technologies that use a VM-like interface like VMs (i.e. [Katacontainers](https://katacontainers.io/)) and [Unikernels](https://en.wikipedia.org/wiki/Unikernel) in general. 
+We note that the content of the blog talks about the internal workings of [runnc](https://github.com/nabla-containers/runnc).
+
+In this blog post, we talk about bridging the network interfaces (pun intended - as will be very apparent later in implementation). More specifically, we will talk about the network interfaces provided by container orchestration systems (usually a `veth` pair), and how we bridged them to what nabla containers consume (a tap device).
+
+
+### Our Kubernetes setup and it's challenges
+
+The nabla containers are based on unikernels, and use the solo5 interface. Solo5 consumes a block device for storage and a tap device for network (since the higher level filesystem and transport layer protocols are part of the LibOS - which is what gives us isolation)! Therefore, the usual way of setting up a nabla process is by creating a tap device and using a static IP or via DHCP.
+
+However, the way networking is done in kubernetes is through a [network plugin](https://kubernetes.io/docs/concepts/cluster-administration/network-plugins/). Most commonly the [Container Network Interface (CNI) plugin](https://kubernetes.io/docs/concepts/cluster-administration/networking/)
+is used to manage the virtualized network setup. The chosen CNI plugin used with most kubernetes clusters provides a [`veth` pair](http://man7.org/linux/man-pages/man4/veth.4.html). Therefore, there is a need to be able to connect the tap device of the unikernel to the interface setup by the CNI plugin. 
+
+It is important to note that there are CNI plugins that provide tap devices. However, we still require system containers to run as regular containers, as they may need to run as privileged host components. It is possible to set it up such that the different types of containers get their own set of interfaces (i.e. regular containers get a `veth` pair and nabla containers get a tap). 
```

Comment:
I think the other (more important) reason for using the veth pair CNI is that it's the default in Docker. Docker is really the setup used by the majority of nabla "users" and developers.

Reply:
Added section:
```
+#### A note on other OCI consumers
+
+Backing up for a moment from the world of kubernetes, we can't forget that the OCI runtime interface has more consumers. One of them being our most popular way to run nabla containers, docker! In docker's case, the default behavior is to provide a `veth` pair!
+
```